### ISSUE TITLE
for await-latest ?

### ISSUE THREAD

**Issue by ursusursus** [apple/swift-async-algorithms#328]
Hi, we need to be able to iterate over async sequence such that if the iteration does some async stuff & upstream happens the emit in the mkddle of it, the current iteration is cancelled & next ine is subscribed (i.e. it doesn't wait)

```
for item in items.latest {
   try await sleep(long time)
}
```

this would mirror kotlin's `collectLatest (also there's flatMapLatest, transformLatest` which are very powerful & necessary

**Comment by twittemb** [comment]
Hi. I did that some time ago in this repo https://github.com/sideeffect-io/AsyncExtensions

Perhaps we could reproduce that in this Apple repo.

**Comment by ursusursus** [comment]
Where exactky please? I only see https://github.com/sideeffect-io/AsyncExtensions/blob/main/Sources/Operators/AsyncSequence%2BCollect.swift

**Comment by ursusursus** [comment]
Well that requires to return a sequence, can it be modified to act as collect(Latest)?

**Comment by EngOmarElsayed** [comment]
Can I work on this issue @twittemb

**Comment by twittemb** [comment]
Sure

**Comment by EngOmarElsayed** [comment]
@twittemb Can you give some explanation for the issue because the above comment is not clear

**Comment by twittemb** [comment]
I think the idea was to mimic operators from reactive programming that switch to the latest upstream of events. 

When you have a stream of streams every time the parent stream emits a new stream then the client switches to this latest stream and receive its events ... the previous subscription being cancelled.

**Comment by EngOmarElsayed** [comment]
@twittemb So basically the idea is to get to the latest stream without having to await for the rest. I hope I understood it correctly

**Comment by twittemb** [comment]
As I said in the comments, I have an implementation here (https://github.com/sideeffect-io/AsyncExtensions/blob/main/Sources/Operators/AsyncSequence+FlatMapLatest.swift) you can have a look if you want.

**Comment by EngOmarElsayed** [comment]
Thanks man @twittemb , I will finish this issue soon

**Comment by ursusursus** [comment]
I don't think that's the same.

I don't want to subscribe another stream, just cancel processing of the current emit and move to processing the latest.

Or is that the same you think?

Here's docs for kotlin's variant
https://kotlinlang.org/api/kotlinx.coroutines/kotlinx-coroutines-core/kotlinx.coroutines.flow/collect-latest.html

As you see the closure `action: suspend (T) -> Unit` it takes, is a) suspend (read `async`), returns unit (read `Void`), so not a Sequence, like in the `flatMapLatest` proposal you've been given

**Comment by EngOmarElsayed** [comment]
So basically the goal is to process the last stream and ignoring the rest, right ?

**Comment by ursusursus** [comment]
Not sure why you say "stream". Isn't stream always the same?

I'd say the goal is to process each "emit" but on backpressure instead of queueing, you cancel the current and move onto the latest

**Comment by lyzkov** [comment]
> As I said in the comments, I have an implementation here (https://github.com/sideeffect-io/AsyncExtensions/blob/main/Sources/Operators/AsyncSequence+FlatMapLatest.swift) you can have a look if you want.

This implementation seems to have potential performance issues. Is there any other implementation of swithToLatest available for AsyncExtensions?

**Comment by EngOmarElsayed** [comment]
what are the performance issues ?

**Comment by lyzkov** [comment]
@EngOmarElsayed I've overlooked/misdiagnosed higher CPU load.

TLDR;

For some unknown reasons, in certain conditions, the CPU load reported by Xcode for my background processing app jumps from ~25% to ~50%. It was happening much more frequently as I introduced configuration complexities to my asynchronous pipelines. One of them was using a custom timeout function that triggered a higher CPU load immediately after starting the processing of the asynchronous sequence.
`flatMapLatest` seemed to contribute to that trigger but after careful investigation, it looks fine.

What can be a reason for such a CPU load jumps in an asynchronous environment?
